src/main/java/org/opensearch/plugin/insights/core/service/QueryInsightsService.java

```diff
@@ -114,15 +121,17 @@ public QueryInsightsService(
         this.searchQueryMetricsEnabled = clusterSettings.get(SEARCH_QUERY_METRICS_ENABLED_SETTING);
         this.searchQueryCategorizer = SearchQueryCategorizer.getInstance(metricsRegistry);
         clusterSettings.addSettingsUpdateConsumer(SEARCH_QUERY_METRICS_ENABLED_SETTING, this::setSearchQueryMetricsEnabled);
+        this.groupingType = DEFAULT_GROUPING_TYPE;
     }
 
     /**
      * Ingest the query data into in-memory stores
      *
      * @param record the record to ingest
+     * @return true/false
      */
     public boolean addRecord(final SearchQueryRecord record) {
-        boolean shouldAdd = searchQueryMetricsEnabled;
+        boolean shouldAdd = isSearchQueryMetricsFeatureEnabled() || isGroupingEnabled();
```

Comment:
This is because if grouping is enabled, we need to add all records to the queryRecordsQueue and avoid the following optimization in QueryInsightsService:
```
// skip add to top N queries store if the incoming record is smaller than the Nth record
                if (currentSnapshot.size() < entry.getValue().getTopNSize()
                    || SearchQueryRecord.compare(record, currentSnapshot.get(0), entry.getKey()) > 0) {
                    shouldAdd = true;
                    break;
                }
```

Let me know if you think otherwise

Reply:
>If only query metrics is enabled we always add the records.
If only top N is enabled we perform an optimization and skip adding the records if they do not make it to the Top N.
If grouping is enabled we need to add all the records since we cannot perform the optimization above.

>Not sure what you are referring to here?

My concern was mostly around the fact that as a user if I disable this framework via setting, we are still adding the records in the background(which are never going to be used) due to grouping being enabled which doesn't look very right to me. I would think less about the optimization/accuracy here compared to correctness.